File: docs/rudymap/rudymap_download.php
#!/usr/bin/php
<?php
/// check version and donwload the weekly release
$admin_email ="happyman";
$fpath = "/home/happyman/mapsforge";
$cpath = "/home/happyman/stache";
$cur_ver_fpath = $fpath . "/VERSION";
$want = array("MOI_OSM_Taiwan_TOPO_Rudy.map.zip","MOI_OSM_Taiwan_TOPO_Rudy.poi.zip","MOI_OSM_Taiwan_TOPO_Rudy_style.zip");
$base = "http://rudy.basecamp.tw/";
//$base = "https://rudy.outdoors.tw/";
// prevent cache
$ver_data = file_get_contents($base . "/index.json". '?'.mt_rand());
$dirty = 0;
$v = json_decode($ver_data, true);

$opt = getopt("v");
if (isset($opt['v']))  
	$version_only = 1; 
else 
	$version_only = 0;

if (file_exists($cur_ver_fpath)) {
	$cur_ver = trim(file_get_contents($cur_ver_fpath));
}
if (empty($cur_ver)) 
{
	$cur_ver = "v0.0";
}

echo "cur_ver = $cur_ver ";
// print_r($v);
foreach($v as $vv) {
	if (in_array($vv['name'], $want)) {
		
		if ($vv['name'] == "MOI_OSM_Taiwan_TOPO_Rudy.map.zip" &&
				$vv['version'] != $cur_ver ) {
				$dirty = 1;
				$ver = $vv['version'];
		}
	}

	
}
if ($dirty == 1 ) {
echo ";online_ver = $ver\n";
	if ($version_only == 1 ) 
		exit(0);
	do_update($base,$ver,$cur_ver);
} else {
	echo "no update required..\n";
	mail($admin_email,"rudy map checked", "version is now $cur_ver");
}
//print_r($v);
function do_update($base,$ver,$old_ver) {
	// 1. download  and unzip 
	global $fpath,$cpath,$admin_email;
	
	@mkdir($fpath . "/$ver", 0755, true);
	chdir($fpath . "/$ver");
	$zips = array("MOI_OSM_Taiwan_TOPO_Rudy.map.zip","MOI_OSM_Taiwan_TOPO_Rudy.poi.zip","MOI_OSM_twmap_style.zip", "MOI_OSM_bn_style.zip","MOI_OSM_dn_style.zip", "MOI_OSM_Taiwan_TOPO_Rudy_style.zip","MOI_OSM_tn_style.zip");
	foreach($zips as $zip) {
		//my_system(sprintf("wget -O %s %s",$zip , $base . $zip));
		my_system(sprintf("rsync -avP homevm:rudy/static/%s ./%s",$zip,$zip));
		my_system("unzip -o $zip");
	}
	// check version
	// strings MOI_OSM_Taiwan_TOPO_Rudy.map
	$cmd=sprintf("strings MOI_OSM_Taiwan_TOPO_Rudy.map |grep -e \"RuMAP %s\"",$ver);
	my_system($cmd);
	chdir($fpath);
       // 3. clean tile cache
	echo "tilestache clean...\n";
	$layers = array("moi_osm","moi_osm_gpx","rudy_default","moi_happyman_nowp_nocache","twmap_happyman_nowp_nocache");
	foreach($layers as $layer) {
		echo "cleaning $layer...\n";
		$old=$layer.".old";
		chdir($cpath);
		if (file_exists($old)) {
			exec("rm -r $old");
			echo "$old removed\n";
		}
		exec("mv $layer $old &&  mkdir $layer");
		echo "$layer created\n";
	}
	// 4. serve it 
	chdir($fpath);
	exec("rm -f cur; ln -s $ver cur");
	// 5. update VERSION
	echo "update VERSION file\n";
	file_put_contents($fpath . "/VERSION", $ver);
	// 6. restart java
	echo "restart java tile server...\n";
	system("killall -9 java");
	// 7. update prop file
	// exec("cp MOI_OSM.xml.prop cur");
	// 8. purgecache from cloudflare
	//exec("bash /home/happyman/bin/purgecache");
	purge_cache();
	// 9. delete unzipped files from old directory
	echo "clean old directory...\n";
	exec("cd $fpath; bash rudymap_clean.sh $old_ver");
	// 10. email me
	mail($admin_email,"rudy map updated!", "version is now $ver");
	

}
function my_system($cmd) {
	global $ver;
	exec($cmd, $out, $ret);
	if ($ret != 0 ){
		exec("rm -r $fpath/$ver");
		echo "cmd fails: $cmd\n";
		exit(1);
	}
}
function purge_cache(){
	$auth_email="";
	$auth_key="";
	if (empty($auth_email)) return;
	$cmd=sprintf('curl -X POST "https://api.cloudflare.com/client/v4/zones/47e7b5f66904a67bbd6a7e58eda91331/purge_cache" -H "X-Auth-Email: %s" -H "X-Auth-Key: %s" -H "Content-Type: application/json" --data \'{"purge_everything":true}\'',$auth_email,$auth_key);
	echo $cmd;
	exec($cmd);
}

File: docs/rudymap/rudymap_download.py
#!/usr/bin/env python3
import json
import os
import random
import subprocess
import sys

from config import (
    CF_API_KEY,
    CF_EMAIL,
    CF_ZONE_ID,
    RUDY_ADMIN_EMAIL as ADMIN_EMAIL,
    RUDY_BASE as BASE,
    RUDY_CUR_VER_FPATH as CUR_VER_FPATH,
    RUDY_FPATH as FPATH,
    RUDY_MAIN_LAYERS as MAIN_LAYERS,
    RUDY_NOCACHE_LAYERS as NOCACHE_LAYERS,
    RUDY_RSYNC_REMOTE as RSYNC_REMOTE,
    RUDY_TILESTACHE_CFG as TILESTACHE_CFG,
    RUDY_TILESTACHE_MOIOSM_CFG as TILESTACHE_MOIOSM_CFG,
    RUDY_WANT as WANT,
    RUDY_ZOOMS as ZOOMS,
)


def main():
    version_only = "-v" in sys.argv

    r = random.randint(0, 999999)
    result = subprocess.run(
        ["curl", "-sL", f"{BASE}/index.json?{r}"], capture_output=True, text=True
    )
    v = json.loads(result.stdout)

    cur_ver = ""
    if os.path.exists(CUR_VER_FPATH):
        with open(CUR_VER_FPATH) as f:
            cur_ver = f.read().strip()
    if not cur_ver:
        cur_ver = "v0.0"

    print(f"cur_ver = {cur_ver} ", end="")

    dirty = False
    online_ver = None
    for vv in v:
        if vv["name"] in WANT:
            if (
                vv["name"] == "MOI_OSM_Taiwan_TOPO_Rudy.map.zip"
                and vv["version"] != cur_ver
            ):
                dirty = True
                online_ver = vv["version"]

    if dirty:
        print(f";online_ver = {online_ver}")
        if version_only:
            sys.exit(0)
        do_update(online_ver, cur_ver)
    else:
        print("no update required..")
        subprocess.run(
            ["mail", "-s", "rudy map checked", ADMIN_EMAIL],
            input=f"version is now {cur_ver}",
            text=True,
        )


def do_update(ver, old_ver):
    ver_dir = os.path.join(FPATH, ver)

    def my_system(cmd):
        result = subprocess.run(cmd)
        if result.returncode != 0:
            subprocess.run(["rm", "-rf", ver_dir])
            print(f"cmd fails: {cmd}")
            sys.exit(1)

    # 1. download and unzip
    os.makedirs(ver_dir, mode=0o755, exist_ok=True)
    os.chdir(ver_dir)

    zips = [
        "MOI_OSM_Taiwan_TOPO_Rudy.map.zip",
        "MOI_OSM_Taiwan_TOPO_Rudy.poi.zip",
        "MOI_OSM_twmap_style.zip",
        "MOI_OSM_bn_style.zip",
        "MOI_OSM_dn_style.zip",
        "MOI_OSM_Taiwan_TOPO_Rudy_style.zip",
        "MOI_OSM_tn_style.zip",
    ]
    for zip_name in zips:
        my_system(["rsync", "-avP", f"{RSYNC_REMOTE}{zip_name}", f"./{zip_name}"])
        my_system(["unzip", "-o", zip_name])

    # 2. verify version
    my_system(
        [
            "bash",
            "-c",
            f"strings MOI_OSM_Taiwan_TOPO_Rudy.map | grep -e 'RuMAP {ver}'",
        ]
    )

    os.chdir(FPATH)

    # 3. flush memcached
    memcached_hosts = ["localhost:11211", "192.168.98.111:11211"]
    for host in memcached_hosts:
        print(f"flushing memcached {host}...")
        result = subprocess.run(
            ["bash", "-c", f"echo 'flush_all' | nc -N {host}"],
            capture_output=True, text=True,
        )
        if result.returncode != 0:
            print(f"warning: memcached flush for {host} failed (exit {result.returncode})")
        elif "OK" not in result.stdout:
            print(f"warning: memcached flush for {host} response: {result.stdout.strip()}")

    # 4. symlink cur -> ver
    os.chdir(FPATH)
    cur_link = os.path.join(FPATH, "cur")
    if os.path.lexists(cur_link):
        os.remove(cur_link)
    os.symlink(ver, cur_link)

    # 5. update VERSION
    print("update VERSION file")
    with open(CUR_VER_FPATH, "w") as f:
        f.write(ver)

    # 6. restart java tile server
    print("restart java tile server...")
    subprocess.run(["killall", "-9", "java"], capture_output=True)

    # 7. purge cloudflare cache
    purge_cache()

    # 8. email
    subprocess.run(
        ["mail", "-s", "rudy map updated!", ADMIN_EMAIL],
        input=f"version is now {ver}",
        text=True,
    )


def purge_cache():
    print("purging cloudflare cache...")
    result = subprocess.run(
        [
            "curl", "-X", "POST",
            f"https://api.cloudflare.com/client/v4/zones/{CF_ZONE_ID}/purge_cache",
            "-H", f"X-Auth-Email: {CF_EMAIL}",
            "-H", f"X-Auth-Key: {CF_API_KEY}",
            "-H", "Content-Type: application/json",
            "--data", '{"purge_everything":true}',
        ],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        print(f"warning: cloudflare purge failed (exit {result.returncode}): {result.stderr}")
    else:
        print(result.stdout)


if __name__ == "__main__":
    main()
